### PR TITLE
Use -fsanitize=fuzzer if clang supports it

### PR DESCRIPTION
**Pull request by evverx**
It's more of an attempt to figure out what to do about `-fsanitize=fuzzer` (which is currently optional on ClusterFuzz but judging by https://github.com/google/oss-fuzz/issues/2317 will be mandatory sooner or later). I also hope I'll finally fix https://github.com/systemd/systemd/issues/10645 (that's been bugging me for no apparent reason :-))

### CODE REVIEW

**Comment by evverx** [comment]
I'm almost done here. https://github.com/systemd/systemd/issues/10645 is fixed and the fuzzers seem to be built successfully with `./infra/helper.py` (with `--engine=libfuzzer` and `--engine=afl`). I hope I didn't break coverage reports (but who needs them anyway, right? :-)).

**Comment by evverx** [comment]
Coverage reports are fine (though they could be better in terms of coverage but unfortunately it has nothing to do with this PR :-)). I'll go ahead and remove the "dont-merge" label because I think the PR is ready to be reviewed.

File: meson.build [hunk 1]
```diff
@@ -31,11 +31,12 @@ substs.set('PROJECT_VERSION',      meson.project_version())
 project_source_root = meson.current_source_dir()
 
 want_ossfuzz = get_option('oss-fuzz')
+want_ossfuzz_without_fuzzing_engine = get_option('oss-fuzz-without-fuzzing-engine')
 want_libfuzzer = get_option('llvm-fuzz')
```
Comment (jonathanmetzman) [inline meson.build:35]: What is want_libfuzzer for? Is it for building the fuzzers with LF outside of OSS-Fuzz?

File: meson.build [hunk 2]
```diff
@@ -300,7 +301,14 @@ else
 endif
 
 if want_libfuzzer
-        fuzzing_engine = meson.get_compiler('cpp').find_library('Fuzzer')
+        fuzzing_engine = meson.get_compiler('cpp').find_library('Fuzzer', required : false)
+        if fuzzing_engine.found()
+                add_project_arguments('-fsanitize-coverage=trace-pc-guard,trace-cmp', language : 'c')
```
Comment (jonathanmetzman) [inline meson.build:306]: Why are you using trace-pc-guard instrumentation with libFuzzer? It's deprecated, you should be using fuzzer-no-link.

File: meson.build [hunk 3]
```diff
@@ -300,7 +301,14 @@ else
 endif
 
 if want_libfuzzer
```
Comment (jonathanmetzman) [inline meson.build:303]: Is want_libfuzzer supposed to be true when want_ossfuzz_without_fuzzing_engine is true?

File: meson.build (hunk 1 above)
Comment (evverx) [reply meson.build:35]: @jonathanmetzman thank you for taking a look! That's correct. `want_libfuzzer` is used outside of OSS-Fuzz to make it easier to build the fuzzers without `./infra/helper.py` and docker. It's also usually much faster to just install `clang` (and optionally `libfuzzer` depending on which version of clang is installed) and run something like `CC=clang-5.0 CXX=clang++-5.0 ./tools/oss-fuzz.sh`.

File: meson.build (hunk 2 above)
Comment (evverx) [reply meson.build:306]: The "want_libfuzzer" mode is supposed to also work on systems where `fuzzer-no-link` isn't supported. Last time I checked clang-3.9, clang-4.0 and clang-5.0 bailed out receiving `-fsanitize=fuzzer-no-link`. And, as far as I know, `libFuzzer` isn't shipped as a library with clang-6+.

File: meson.build (hunk 3 above)
Comment (evverx) [reply meson.build:303]: `want_libfuzzer` and `want_ossfuzz*` can't be used together. `meson` will bail out if both `want_libfuzzer` and `want_ossfuzz` are set to `true`.

File: meson.build (hunk 2 above)
Comment (jonathanmetzman) [reply meson.build:306]: Ah OK, yeah it should still work with old versions of LF.

>And, as far as I know, libFuzzer isn't shipped as a library with clang-6+.

What do you mean by this?

Comment (evverx) [reply meson.build:306]: > What do you mean by this?

While I was answering the question about why I use `trace-pc-guard` I realized that I can add another `if` to check whether `fuzzer-no-link` is supported but then remembered that it's very unlikely that `libFuzzer` is installed and `fuzzer-no-link` is supported at the same time.

**Comment by evverx** [comment]
@jonathanmetzman I've just seen an MSan report that looks like the one you mentioned in https://github.com/google/oss-fuzz/issues/2332 (after I pressed CTRL-C during `./infra/helper.py run_fuzzer`). Should I put this PR on hold until the issue is resolved?

In case it's helpful below is the backtrace:
```sh
^C==13== libFuzzer: run interrupted; exiting
Uninitialized bytes in __interceptor_opendir at offset 12 inside [0x7030000224a0, 26)
==13==WARNING: MemorySanitizer: use-of-uninitialized-value
==13==WARNING: Can't read from symbolizer at fd 3
    #0 0x46fdb3 in fuzzer::IterateDirRecursive(std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> > const&, void (*)(std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> > const&), void (*)(std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> > const&), void (*)(std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> > const&)) /src/llvm/projects/compiler-rt/lib/fuzzer/FuzzerIOPosix.cpp:87:12
    #1 0x471666 in fuzzer::Fuzzer::InterruptCallback() /src/llvm/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:269:3
    #2 0x47161f in fuzzer::Fuzzer::StaticInterruptCallback() /src/llvm/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:215:6
    #3 0x52c3fb in SignalAction(int, void*, void*) /src/llvm/projects/compiler-rt/lib/msan/msan_interceptors.cc:1004:3
    #4 0x7fc87b16938f  (/lib/x86_64-linux-gnu/libpthread.so.0+0x1138f)
    #5 0x7fc87ba54aaf  (/out/src/shared/libsystemd-shared-242.so+0x6dfaaf)
    #6 0x7fc87ba1effa in sd_netlink_attach_event /work/build/../../src/systemd/src/libsystemd/sd-netlink/sd-netlink.c:791:13
    #7 0x54b90a in manager_connect_rtnl /work/build/../../src/systemd/src/network/networkd-manager.c:960:13
    #8 0x54ab9f in manager_new /work/build/../../src/systemd/src/network/networkd-manager.c:1393:13
    #9 0x53940b in LLVMFuzzerTestOneInput /work/build/../../src/systemd/src/network/fuzz-network-parser.c:25:9
    #10 0x472914 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:529:15
    #11 0x472305 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) /src/llvm/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:453:3
    #12 0x473b58 in fuzzer::Fuzzer::MutateAndTestOne() /src/llvm/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:669:19
    #13 0x4749b5 in fuzzer::Fuzzer::Loop(std::Fuzzer::vector<std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> >, fuzzer::fuzzer_allocator<std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> > > > const&, std::Fuzzer::vector<std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> >, fuzzer::fuzzer_allocator<std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> > > > const&) /src/llvm/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:814:5
    #14 0x465c1a in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm/projects/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:776:6
    #15 0x48c112 in main /src/llvm/projects/compiler-rt/lib/fuzzer/FuzzerMain.cpp:19:10
    #16 0x7fc87a48382f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #17 0x427138 in _start (/out/fuzz-network-parser+0x427138)

  Uninitialized value was created by an allocation of 'netdev' in the stack frame of function 'network_resolve_stacked_netdevs'
    #0 0x55f1f0 in network_resolve_stacked_netdevs /work/build/../../src/systemd/src/network/networkd-network.c:115

SUMMARY: MemorySanitizer: use-of-uninitialized-value /src/llvm/projects/compiler-rt/lib/fuzzer/FuzzerIOPosix.cpp:87:12 in fuzzer::IterateDirRecursive(std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> > const&, void (*)(std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> > const&), void (*)(std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> > const&), void (*)(std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> > const&))
```

**Comment by evverx** [comment]
Just to clarify, this PR is more or less ready to roll. I added the "dont-merge" label because I'm waiting for https://github.com/google/oss-fuzz/issues/2332 to be fixed. Judging by https://github.com/google/oss-fuzz/issues/2332#issuecomment-486891631, it should happen soon, as far as I can tell.

**Comment by evverx** [comment]
I downloaded the latest images (with clang r359254) and I can't reproduce the issue any more. Looks like it's time to remove the "dont-merge" label.

**Comment by jonathanmetzman** [comment]
Yep. I was planning on waiting a bit to make sure the roll sticks but I'm pretty sure it will at this point.

**Comment by evverx** [comment]
To judge from https://github.com/google/oss-fuzz/issues/2369, it seems to be risky to switch to `-fsanitize=fuzzer`. I'm not sure how likely it is our fuzzers will hit it though.

**Comment by evverx** [comment]
Since I've been actively using FuzzBuzz in the past three days (which allows me to experiment with our fuzz targets before they make it to the repository: [1](https://github.com/systemd/systemd/pull/11528#issuecomment-489453464), [2](https://github.com/systemd/systemd/pull/12230#issuecomment-489450783)), I started to work on https://github.com/systemd/systemd/pull/11887, which conflicts with this PR, again. I'll add the "dont-merge" label here and get back to this PR as soon as I'm done there.

**Comment by evverx** [comment]
I'm going to put this PR on hold. In case anyone gets round to it, it's worth mentioning that without `libFuzzingEngine` `meson` [stops using `$CPP` as a linker](https://github.com/google/oss-fuzz/blob/master/docs/new_project_guide.md#requirements) so $CPP should be pinned down with `link_language` or something like that.

**Pull request closed without merging**